src/Client.php

```diff
@@ -41,12 +47,19 @@ final class Client
     /**
      * Instantiate a new Client.
      *
-     * @param string          $endpoint
-     * @param array           $options
-     * @param HttpClient|null $httpClient
+     * @param ErrorResponseHandler $errorResponseHandler
+     * @param string               $endpoint
+     * @param array                $options
+     * @param HttpClient|null      $httpClient
      */
-    public function __construct(string $endpoint, array $options = [], HttpClient $httpClient = null)
-    {
+    public function __construct(
+        ErrorResponseHandler $errorResponseHandler,
```

Comment:
Finally, I'm not sure it worth it to set it on the constructor. It's a single class and should not be replaced by something else.

Having a separated class for this is good to not make the client to heavy, but it should be instantiated directly into the constructor and not as a required parameter.

What do you think?

Reply:
I feel it gives a better separation of concerns, the client is already over 200 lines of code now, so offloading some functionality to its own class is providing clarity and maintainability, following principles like SOLID. Also this allows for testing response handling in isolation of the rest of the process.

As far as it being heavy or not, Symfony will instantiate both classes pretty much at the same time and inject the ErrorResponseHandler into the client (unless defined as a lazy loaded service), I'm guessing that will give a similar overhead as to instantiating it in the constructor.
In any case the difference is probably not significant enough to state that one way is better than the other. 

However instantiating it in the constructor would take away the ability to test the classes in isolation. I think what you're looking for is something like an inner or friend class, but PHP doesn't have such class scopes.

For standalone usage of this package it would convenient not to have to pass the error handler to the client on each instantiation, we could add a Factory for that?

Reply:
> I feel it gives a better separation of concerns, the client is already over 200 lines of code now, so offloading some functionality to its own class is providing clarity and maintainability, following principles like SOLID

We agree about that, no problem. :+1: 

> However instantiating it in the constructor would take away the ability to test the classes in isolation.

I'm not sure to understand here. When you test the `Client`, you always have to passe a handler instance to it, so what is the difference instanciating it on the constructor?

Reply:
The difference is that you are not able to pass a mock object instead of the real thing that way. Thus to reach 100% code coverage you would have to test the logic of the error handler together with the client, the tests will become more complex and perhaps even more fragile because of that.

Also I can imagine that at the moment this library is used for just the webhook API, but in the future you might want to support the other APIs as well, which will produce different errors, requiring a different error handler.

Anyway it is up to you, but we'd have to refactor the ClientUnitTest and instead of mocking the error handler we'd have to mock the response objects to make sure that the error handler does the right thing for each test.

Reply:
We could use this, to overcome the isolation problem btw:
http://docs.mockery.io/en/latest/reference/creating_test_doubles.html#overloading

But it gets a bit tricky (fragile) when you need to retain the global state and need the error handler in a different test as well..

Reply:
> The difference is that you are not able to pass a mock object instead of the real thing that way.

This is true, but according to the concerning object, is it really worth it to mock it? If not, I don't think it worth it to force it as a constructor argument.